### PR TITLE
Ensures guava callback to complete chained futures

### PR DESCRIPTION
**Pull request by RustedBones**
- Catch all `Throwable` on Guava callbacks to ensure returned future is completed
- Move from `computeIfAbsent` to `put` or `putIfAbsent` since keys are UUIDs and lazy computation was useless
- Simplify semaphore and cache handling in `BaseAsyncLookupDoFn`
- Reset `inFlightRequests` and `semaphore` permits in `startBundle` to avoid leaks/locks
- Apply same code structure on `BaseAsyncDoFn`

### CODE REVIEW

File: scio-core/src/main/java/com/spotify/scio/transforms/FutureHandlers.java [hunk 1]
```diff
@@ -92,7 +99,7 @@ default void waitForFutures(Iterable<CompletableFuture<V>> futures)
         throws InterruptedException, ExecutionException {
       CompletableFuture[] array =
           StreamSupport.stream(futures.spliterator(), false).toArray(CompletableFuture[]::new);
-      CompletableFuture.allOf(array).get();
```
Comment (RustedBones) [inline scio-core/src/main/java/com/spotify/scio/transforms/FutureHandlers.java:95]: all implementation do not throw in `waitForFutures` but this one. Change to match same behavior

File: scio-test/src/test/scala/com/spotify/scio/transforms/AsyncLookupDoFnTest.scala [hunk 2]
```diff
@@ -65,8 +64,9 @@ class AsyncLookupDoFnTest extends PipelineSpec {
   )(tryFn: T => Try[String]): Unit = {
     val output = runWithData(1 to 10)(_.parDo(doFn)).map { kv =>
       val r = tryFn(kv.getValue) match {
-        case Success(v) => v
-        case Failure(e) => e.getMessage
+        case Success(v)                      => v
+        case Failure(e: CompletionException) => e.getCause.getMessage
```
Comment (RustedBones) [inline scio-test/src/test/scala/com/spotify/scio/transforms/AsyncLookupDoFnTest.scala:68]: Java future wraps the exception in a `CompletionException`. Unwrap for the test

File: scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncLookupDoFn.java [hunk 3]
```diff
@@ -246,6 +203,7 @@ public void finishBundle(FinishBundleContext context) {
         throw new RuntimeException("Failed to process futures", e);
       } catch (ExecutionException e) {
         LOG.error("Failed to process futures", e);
+        throw new RuntimeException("Failed to process futures", e);
```
Comment (RustedBones) [inline scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncLookupDoFn.java:206]: This was probably removed due to inconsistent behaviour of the java future, throwing in `waitForFutures` if a any of the future failed.

File: scio-test/src/test/scala/com/spotify/scio/transforms/FutureHandlersTest.scala [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        handler.waitForFutures(List[F](f1, f2).asJava)
+      }
+
+      fail(f1)(new Exception("f1 failed"))
+      backgroundTask.isCompleted shouldBe false
+      complete(f2)("f2 done")
+      noException shouldBe thrownBy {
+        Await.result(backgroundTask, 1.second)
+      }
+    }
```
Comment (RustedBones) [inline scio-test/src/test/scala/com/spotify/scio/transforms/FutureHandlersTest.scala:52]: This test was failing for java

File: scio-test/src/test/scala/com/spotify/scio/transforms/FutureHandlersTest.scala [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+      )
+      val e = new Exception("failure")
+      fail(f)(e)
+      result shouldBe Failure(e)
+      val ee = the[ExecutionException] thrownBy (access(chainedFuture))
+      val cause = ee.getCause
+      cause.getMessage shouldBe "failure"
+      cause.getSuppressed.headOption.map(_.getMessage) shouldBe Some("callback failure")
+    }
+  }
```
Comment (RustedBones) [inline scio-test/src/test/scala/com/spotify/scio/transforms/FutureHandlersTest.scala:134]: Those tests were failing for guava, propagating the exception in the callback thread (very probably from client implementation)

**Review by RustedBones — COMMENTED** [review]
@farzad-sedghi @shnapz  Added unit test as discussed

File: scio-core/src/main/java/com/spotify/scio/transforms/FutureHandlers.java [hunk 6]
```diff
@@ -45,7 +45,7 @@ public class FutureHandlers {
     default void waitForFutures(Iterable<ListenableFuture<V>> futures)
         throws InterruptedException, ExecutionException {
       // Futures#allAsList only works if all futures succeed
-      Futures.whenAllComplete(futures).run(() -> {}, MoreExecutors.directExecutor()).get();
+      Futures.successfulAsList(futures).get();
```
Comment (clairemcginty) [inline scio-core/src/main/java/com/spotify/scio/transforms/FutureHandlers.java:48]: just wondering, does this use the same executor (`MoreExecutors.directExecutor()`)? 

Also, how does `successfulAsList` handle failed futures? does it await all futures and only return failed ones?

**Review by clairemcginty — APPROVED** [review]

**Pull request merged**